### PR TITLE
[infra] Use ESM

### PR DESCRIPTION
**Pull request by jablko**
Upgrade the scripts from CommonJS to ESM.

What I did was:
- Replace `require()` with `import`.
- Replace `__dirname` with `import.meta.url`.

### CODE REVIEW

**Comment by jakebailey** [comment]
I'm not sure that this is a good idea to merge; the scripts are only going to run in node anyway and this would make DT harder to contribute to for those using older versions of node that don't support all of these new resolution features.

**Comment by jablko** [comment]
@jakebailey Thanks for reviewing! Does it change the versions of node that the scripts can run under? Already they run under v12.20.0 and above, I think, which has these features?

**Comment by jakebailey** [comment]
I guess the top level DT package is already `"module"`, but I can't recall if node 12 still needs `--experimental-modules` or not to actually enable the feature.

**Comment by jablko** [comment]
@jakebailey I double checked:
```sh
$ wget https://nodejs.org/download/release/v12.20.0/node-v12.20.0-linux-x64.tar.xz
$ tar --extract --file node-v12.20.0-linux-x64.tar.xz 
$ node-v12.20.0-linux-x64/bin/node scripts/not-needed.js 
Usage: npm run not-needed -- typingsPackageName asOfVersion [libraryName]
# ^ Doesn't need the --experimental-modules flag to run the scripts.

File: .gitignore [hunk 1]
```diff
@@ -4,7 +4,6 @@
 *.pdb
 *.suo
 *.js
-*.cjs
```
Comment (jakebailey) [inline .gitignore:7]: Probably best to keep this, so nobody accidentally checks in random .cjs files.

File: scripts/fix-tslint.ts [hunk 2]
```diff
@@ -2,44 +2,44 @@
...
-    const target = path.join(dir, 'tslint.json');
+function fixTslint(dir: URL): void {
+    const target = new URL('tslint.json', dir);
     if (!fs.existsSync(target)) return;
     let json = JSON.parse(fs.readFileSync(target, 'utf-8'));
     json = fix(json);
-    const text = Object.keys(json).length === 1 ? '{ "extends": "@definitelytyped/dtslint/dt.json" }' : JSON.stringify(json, undefined, 4);
-    fs.writeFileSync(target, text + "\n", "utf-8");
+    const text =
+        Object.keys(json).length === 1 ? '{ "extends": "dtslint/dt.json" }' : JSON.stringify(json, undefined, 4);
```
Comment (jakebailey) [inline scripts/fix-tslint.ts:34]: I don't think this extends config should be changing.

File: scripts/update-config/LintPackage.ts [hunk 3]
```diff
@@ -1,8 +1,8 @@
-import * as fs from "fs";
+import * as fs from "node:fs";
 import * as stringify from "json-stable-stringify";
-import * as path from "path";
+import * as path from "node:path";
 import { Configuration as Config, ILinterOptions, Linter, LintResult } from "tslint";
-import * as ts from "typescript";
+import ts from "typescript";
```
Comment (jakebailey) [inline scripts/update-config/LintPackage.ts:5]: I'm a little surprised by this; it's not a `*` import?

File: scripts/update-config/index.ts [hunk 4]
```diff
@@ -56,12 +58,9 @@ async function main() {
 }
 
 function dtConfig(updatedRules: string[]): Config.IConfigurationFile {
-    const resolvedDtslint = require.resolve('dtslint');
+    const resolvedDtslint = require.resolve("dtslint");
```
Comment (jakebailey) [inline scripts/update-config/index.ts:61]: I'm a little confused by the quote styles used in this PR, some files change to single quotes, while some change to double quotes. Is it just to be consistent within the file? There are other formatting changes that would imply `prettier` or something, but if so, we should end up with consistent quoting.

**Review by jakebailey — CHANGES_REQUESTED** [review]
Seems OK with some comments. I'd like someone on the TS team other than me look at this to make sure there's nothing I'm missing about making a change like this.

File: .gitignore (hunk 1 above)
Comment (jablko) [reply .gitignore]: - [x] I've removed this hunk.

File: scripts/fix-tslint.ts (hunk 2 above)
Comment (jablko) [reply scripts/fix-tslint.ts]: - [x] Thank you! I've undone that change.

File: scripts/update-config/LintPackage.ts (hunk 3 above)
Comment (jablko) [reply scripts/update-config/LintPackage.ts:5]: I double checked:
```sh
$ echo '
import a from "typescript";
a.createSourceFile("input.ts", "");  // ✔️

import * as b from "typescript";
b.createSourceFile("input.ts", "");  // ❌
' > script.js
$ node script.js
file:///script.js:6
b.createSourceFile("input.ts", "");  // ❌
  ^

TypeError: b.createSourceFile is not a function
    at file:///script.js:6:3
    at ModuleJob.run (internal/modules/esm/module_job.js:145:37)
    at async Loader.import (internal/modules/esm/loader.js:182:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5)

Comment (jakebailey) [reply scripts/update-config/LintPackage.ts:5]: Oh, the tsconfig doesn't enable esModuleInterop, maybe that's what's going on. Maybe that's taboo in ESM mode?

Comment (jakebailey) [reply scripts/update-config/LintPackage.ts:5]: Hm, but you did test in js.

File: scripts/fix-tslint.ts [hunk 5]
```diff
@@ -2,44 +2,46 @@
...
 /// <reference types="node" />
 
-import * as fs from "fs";
-import * as path from "path";
-import JSON = require("comment-json");
+import * as fs from 'node:fs';
+import JSON from 'comment-json';
 
-const home = path.join(__dirname, "..", "types");
+const home = new URL('../types/', import.meta.url);
```
Comment (sandersn) [inline scripts/fix-tslint.ts:8]: why change from `path.join` to `new URL`?

File: scripts/fix-tslint.ts [hunk 6]
```diff
@@ -2,44 +2,46 @@
...
 
 for (const dirName of fs.readdirSync(home)) {
-    if (dirName.startsWith(".") || dirName === "node_modules" || dirName === "scripts") {
+    if (dirName.startsWith('.') || dirName === 'node_modules' || dirName === 'scripts') {
```
Comment (sandersn) [inline scripts/fix-tslint.ts:11]: formatting changes should be in a separate PR, preferably one that sets up prettier to enforce future formatting.

**Review by sandersn — CHANGES_REQUESTED** [review]
This is OK as long as whoever merges it watches DT infrastructure for a week or two afterward. I'll do that if it's this week since it's my DT week.

Also, please undo all formatting and non-essential code changes. I really want it to be just changes to imports and file extensions.

File: scripts/fix-tslint.ts (hunk 5 above)
Comment (jablko) [reply scripts/fix-tslint.ts:8]: `import.meta.url` is a URL, which `fs.readdirSync()` etc. accept but `path.join()` doesn't.

**Review by sandersn — APPROVED** [review]
This change looks better to me now, but I'm swamped this week and can't spare time to watch the DT infrastructure for problems. @weswigham can you merge this next week when you're on DT duty?

**Review by jakebailey — APPROVED** [review]

**Pull request merged**